File: frontend/src/api/types/chain.ts
/**
 * Chains are domain entities, not hardcoded branches. Adding a chain is a
 * matter of metadata; it must never require touching onboarding logic.
 *
 * PROVISIONAL CONTRACT LAYER — see `src/api/types/index.ts`.
 */

/** Chains supported by the MVP. TRON is the chain; TRX is its native token. */
export const KnownChainId = {
  ETHEREUM: 'ethereum',
  BITCOIN: 'bitcoin',
  BNB: 'bnb',
  SOLANA: 'solana',
  LITECOIN: 'litecoin',
  XRPL: 'xrpl',
  TRON: 'tron',
  DOGECOIN: 'dogecoin',
} as const
export type KnownChainId = (typeof KnownChainId)[keyof typeof KnownChainId]

/** Open union so a chain added by the backend does not break the client. */
export type ChainId = KnownChainId | (string & {})

export interface Chain {
  id: ChainId
  /** Human-readable chain name, e.g. "XRP Ledger". */
  name: string
  /** Native token symbol, e.g. `TRX` for TRON. */
  native_asset_symbol: string
  is_supported: boolean
}

File: backend/internal/domain/chain/chain.go
// Package chain models supported blockchain networks. Chains are domain
// entities backed by configuration, never hardcoded branches inside portfolio
// algorithms (§20).
package chain

import (
	"time"

	"github.com/google/uuid"
)

// ID is the stable machine identifier of a chain. Display names are never used
// as identifiers (§21).
//
// The codes below are the wire values the frontend already sends. The
// specification's §21 examples spell two of them differently (bnb_chain,
// xrp_ledger); the divergence and its resolution are recorded in
// openapi/DECISIONS.md.
type ID string

const (
	Ethereum  ID = "ethereum"
	Bitcoin   ID = "bitcoin"
	BNBChain  ID = "bnb"
	Solana    ID = "solana"
	Litecoin  ID = "litecoin"
	XRPLedger ID = "xrpl"
	Tron      ID = "tron"
	Dogecoin  ID = "dogecoin"
)

// Supported lists the MVP chains (§20). TRON is the blockchain; TRX is its
// native asset.
var Supported = []ID{Ethereum, Bitcoin, BNBChain, Solana, Litecoin, XRPLedger, Tron, Dogecoin}

// IsSupported reports whether id is one of the MVP chains.
func IsSupported(id ID) bool {
	for _, supported := range Supported {
		if supported == id {
			return true
		}
	}
	return false
}

// Chain is a supported blockchain network.
type Chain struct {
	ID   ID
	Name string
	// NativeAssetID references the chain's native asset, which has no contract
	// address (§31).
	NativeAssetID uuid.UUID
	// AddressFormat selects the chain-specific validation and normalization
	// rules. A single generic regex across all chains is explicitly wrong (§189).
	AddressFormat AddressFormat
	IsSupported   bool
	CreatedAt     time.Time
	UpdatedAt     time.Time
}

// AddressFormat identifies a family of address encoding rules shared by
// several chains.
type AddressFormat string

const (
	// AddressFormatEVM covers 0x-prefixed 20-byte addresses.
	AddressFormatEVM AddressFormat = "EVM"
	// AddressFormatBitcoinLike covers Base58Check and Bech32 UTXO addresses.
	AddressFormatBitcoinLike AddressFormat = "BITCOIN_LIKE"
	// AddressFormatSolana covers Base58 Ed25519 public keys.
	AddressFormatSolana AddressFormat = "SOLANA"
	// AddressFormatTron covers Base58Check addresses starting with T.
	AddressFormatTron AddressFormat = "TRON"
	// AddressFormatXRPL covers classic XRP Ledger account addresses.
	AddressFormatXRPL AddressFormat = "XRPL"
)
